src/base/filesystemwatcher.cpp

```diff
@@ -145,20 +145,21 @@ void FileSystemWatcher::processPartialTorrents()
     QStringList noLongerPartial;
 
     // Check which torrents are still partial
-    foreach (const QString &torrentPath, m_partialTorrents.keys()) {
+    for (auto i = m_partialTorrents.begin(); i != m_partialTorrents.end(); ) {
```

Comment:
Seems you forgot to do `++i`?

Personally I think this 'idiom' (removing element from containers) is idiotic, it leaves a big hole for so much human errors.
So for this kind of specific usage I would prefer the old code, but before jumping to conclusion, let's hear if others think the same.

Reply:
@Chocobo1 No I didn't.
The code now is:
```cpp
        for (auto i = m_categories.begin(); i != m_categories.end(); ) {
            const QString category = i.key();
            if (category.startsWith(test)) {
                i = m_categories.erase(i);
                result = true;
                emit categoryRemoved(category);
            } else {
                ++i;
            }
        }
```
Looking at the diff's makes it hard to understand. But seeing the code all at once became self-evident.

There's no error here. I think it's a question of practice. I have a embedded programming and find this idiom natural.

Reply:
>The code now is:

Wrong place, here: https://github.com/luis-pereira/qBittorrent/blob/e08cca3377d641cfafb3c023026fefd9f1029348/src/base/filesystemwatcher.cpp#L148

>There's no error here. I think it's a question of practice. I have a embedded programming and find this idiom natural.

I shall not debate over this and just talking from my experience.

Reply:
@Chocobo1 I will also not discuss it.

Reply:
> I was just talking from my experience.

I believe it's pretty evident that even an experienced dev accustomed to the idiom can make mistakes over it easily...

@glassez 
Please speak up if you have strong preferences.
https://github.com/qbittorrent/qBittorrent/pull/8558#discussion_r173372233

Reply:
Yes, But I would write the code in another way.

Reply:
>Please speak up if you have strong preferences.

I'm not sure I understood exactly what it was about...

>Personally I think this 'idiom' (removing element from containers) is idiotic, it leaves a big hole for so much human errors.
So for this kind of specific usage I would prefer the old code, but before jumping to conclusion, let's hear if others think the same.

The second sentence contradicts the first, because the old code also removed items from the container.

Reply:
> I'm not sure I understood exactly what it was about...

In short, the new code uses an explicit iterator which you need to carefully maintain it every time you remove a element, the old code don't. The maintenance is so tedious that one might screw up easily. Not worth the performance gained IMO.

Reply:
We must not optimize at the expense of maintainability. There must be some balance in this. It is better to sacrifice some performance than to get elusive bugs.

As for this particular case, I need measure it first. Maybe there are compromise solutions?

Reply:
The performance gain depends on the torrent list. `FileSystemWatcher` in my view a class where optimizations make sense. I didn't measured although. 

Do you guys find this  less error prone ?
``` cpp

    // Check which torrents are still partial
    auto i = m_partialTorrents.begin();
    while (i != m_partialTorrents.end()) {
        const QString torrentPath = i.key();
        if (!QFile::exists(torrentPath)) {
            i = m_partialTorrents.erase(i);
            continue;
        }
        if (BitTorrent::TorrentInfo::loadFromFile(torrentPath).isValid()) {
            noLongerPartial << torrentPath;
            i = m_partialTorrents.erase(i);
            continue;
        }
        if (i.value() >= MAX_PARTIAL_RETRIES) {
            QFile::rename(torrentPath, torrentPath + ".invalid");
            i = m_partialTorrents.erase(i);
            continue;
        }

        ++(i.value());
        ++i;
    }
```

Reply:
If performance is required in this case, I would reduce iterator interaction points:

``` cpp
    // Check which torrents are still partial
    auto i = m_partialTorrents.begin();
    while (i != m_partialTorrents.end()) {
        const QString torrentPath = i.key();
        bool stillPartial = false;
        if (BitTorrent::TorrentInfo::loadFromFile(torrentPath).isValid()) {
            noLongerPartial << torrentPath;
        }
        else if (i.value() >= MAX_PARTIAL_RETRIES) {
            QFile::rename(torrentPath, torrentPath + ".invalid");
        }
        else if (QFile::exists(torrentPath)) {
            stillPartial = true;
            ++(i.value());
        }

        i = (stillPartial ? i + 1 : partialTorrents.erase(i));
    }
```

Reply:
>The performance gain depends on the torrent list. 

IMO, performance gain *could* be negligible for normal size workload, `QHash::keys()` runs in O(n) and our loop runs in O(n), every lookup in QHash is amortized O(1), in the end you still get O(n) runtime and O(n) space.

Another point is the old code is similar to what functional programming will do, most objects are immutable, the result is you get less bugs slip in and less headache when you need to modify it later.

Reply:
I tested it in different ways. Well, file operations neutralize all possible performance gains.

Reply:
>Well, file operations neutralize all possible performance gains.

But the above applies only to similar cases. In other cases, we can obtain at least a two-fold performance boost, and it deserves attention. We just have to cut the risk. This can be done by implementing some common algorithm (we can't use std algorithms with Qt dictionaries since its iterator has different behavior):
```c++
template <typename Dictionary, typename BinaryPredicate>
void removeIf(Dictionary &dict, BinaryPredicate p)
{
    auto it = dict.begin();
    while (it != dict.end()) {
        it = (p(it.key(), it.value()) ? dict.erase(it) : it + 1);
    }
}
```
Then the code under discussion might look like this:
```c++
removeIf(m_partialTorrents, [&noLongerPartial](const QString &torrentPath, int &value)
{
    if (!QFile::exists(torrentPath))
        return true;

    if (BitTorrent::TorrentInfo::loadFromFile(torrentPath).isValid()) {
        noLongerPartial << torrentPath;
        return true;
    }

    if (value >= MAX_PARTIAL_RETRIES) {
        QFile::rename(torrentPath, torrentPath + ".invalid");
        return true;
    }

    ++value;
    return false;
});
```

Reply:
The `removeIf()` sounds good +1.

Reply:
Or `Utils::Dict::removeIf`?
What is the best place for it? "base/algorithm.h" or "base/utils/dict.h"?

Reply:
>What is the best place for it? "base/algorithm.h" or "base/utils/dict.h"?

"base/algorithm.h", because the filename is more generic, we can have many similar helper functions in it.

Reply:
base/algorithm.h and `Utils::Dict::removeIf` ?

Reply:
>base/algorithm.h and Utils::Dict::removeIf ?

At least, without Utils namespace.
@Chocobo1, what do you prefer, dictRemoveIf in global namespace or removeIf in Dict namespace?

Reply:
I think removeIf in Dict (or QtDict/QDict?) namespace.

Reply:
>or QtDict/QDict?

No, let's leave such prefixes for Qt itself.

@luis-pereira, now you know how to finish it.